src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -20,11 +21,10 @@ internal class SslStreamInternal
         private static readonly AsyncProtocolCallback s_resumeAsyncReadCallback = new AsyncProtocolCallback(ResumeAsyncReadCallback);
         private static readonly AsyncProtocolCallback s_readHeaderCallback = new AsyncProtocolCallback(ReadHeaderCallback);
         private static readonly AsyncProtocolCallback s_readFrameCallback = new AsyncProtocolCallback(ReadFrameCallback);
+        private static readonly Action<Task, object> s_freeWriteBufferCallback = FreeWriteBuffer;
 
-        private const int PinnableReadBufferSize = 4096 * 4 + 32;         // We read in 16K chunks + headers.
-        private static PinnableBufferCache s_PinnableReadBufferCache = new PinnableBufferCache("System.Net.SslStream", PinnableReadBufferSize);
-        private const int PinnableWriteBufferSize = 4096 + 1024;        // We write in 4K chunks + encryption overhead.
-        private static PinnableBufferCache s_PinnableWriteBufferCache = new PinnableBufferCache("System.Net.SslStream", PinnableWriteBufferSize);
+        private const int FrameOverhead = 32;
+        private const int ReadBufferSize = 4096 * 4 + FrameOverhead;         // We read in 16K chunks + headers.
```

Comment:
Where is ReadBufferSize used now?

Reply:
It's not but... I have just been prototyping exactly that. Because we say "min size 5" and fill that and then min size "frame" and then copy/move and carry on now. Interestingly it didn't hurt perf at all. But it causes a move back to the bad old days of reading 5 then frame.

I am thinking that whole ensure logic can go out the door and just use the readframe size. The TLS spec is pretty clear 14^2 for the plain text and then headers etc. TLS 1.3 is exactly the same.

So I say make it 14^2 + headers with a reference and ditch all the resizing. Just the reset to offset zero after a frame is used.

Reply:
And free of course if all data is consumed.